CIP: Ceramic API

oed opened the pull request:
This is an early draft of what could become the Ceramic API. An API that enables developers to read and write events directly to event streams.

### [Discussion forum thread](https://forum.ceramic.network/t/cip-137-ceramic-api/1153)

AaronGoldman commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,200 @@
+
+### `ceramic_unsubscribe`
+
+This method tells the Ceramic node to unsubscribe from a given stream set.
```
If there are multiple clients that want the same stream, do we want something like reference counting.

e.g.
If each subscription was given a UUID and we tracked the controller that asked for it a controller could delete their subscription without deleting the other subscriptions for the same range.

1) alice subscribes to `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`
2) bob subscribes to `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`
3) bob unsubscribes to `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`

at the end of 3 should the node be subscribed to `kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`?

having the controller and UUID would give the unsubscriber the option to delete only their unsubscription or all unsubscriptions for `kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`.

AaronGoldman commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,200 @@
+
+### `ceramic_getRawEvents`
+
+Get the raw IPLD data of a set of events given an array of eventids.
```
This is under specified about recursion.
If the event has a data link instead of a data Json should you pull that block also?
How deep should you go?

AaronGoldman commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,200 @@
+
+---
+
+### `ceramic_addRawEvent`
```
would prefer `put` to imply the idempotency that http means by `put`

AaronGoldman commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,200 @@
+
+## Motivation
+<!--Motivation is critical for CIPs that want to change the Ceramic protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the CIP solves. CIP submissions without sufficient motivation may be rejected outright.-->
+TODO
```
In order to enable databases other than ComposeDB to be built on top of ceramic we need a ceramic API for ceramic event streaming system to communicate with databases.

oed commented:
@AaronGoldman I addressed some of your comments, but can we please keep the normative discussions to the [forum post](https://forum.ceramic.network/t/cip-137-ceramic-api/1153)?

oed commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,200 @@
+
+
+
+### `ceramic_subscribe`
```
So what you are subscribing to is a separator-key/value, in the example it happens to be a model. It can definitely be possible to subscribe to specific streams within a separator!

qbig replied:
Let's say stream-bar is from model-bar. So the "topic" could be "model-foo" or "stream-bar" or explicitly "model-foo/stream-bar"?

AaronGoldman commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,203 @@
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+This CIP introduces a JSON-RPC interface that can be used to subscribe to a number of streams inside of an interest range, and write new events to these streams. The API exposes two main ways of retrieving events. A simple method that uses the nodes local ordering of events to retrieve events by simply keeping track of a single vector client side. More advanced users can use a simplified version  ...
```
why JSON-RPC over say GraphQL

oed replied:
Sure, happy to change this if we have a good reason for it! Let's continue this discussion in the forum

AaronGoldman commented on CIPs/cip-137.md:
```diff
@@ -0,0 +1,203 @@
+
+**Returns:**
+
+- *events* <string> - a base64 encoded CAR file containing the events
```
events' blocks

oed replied:
```suggestion
- *events* <string> - a base64 encoded CAR file containing the event blocks
```
